certora/specs/SocialRecoveryModule.spec

```diff
@@ -1,26 +1,160 @@
+
+    assert safeContract.isModuleEnabled(currentContract);
+    assert e.msg.sender == safeContract;
+    assert currentContract.isGuardian(safeContract, guardian);
+    assert otherAccountIsGuardian == currentContract.isGuardian(safeContract, otherAccount);
+    assert currentGuardiansCount + 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
+}
+
+// This integrity rule verifies that the guardian can always be added considering ideal conditions.
+rule guardianCanAlwaysBeAdded(env e, address guardian, uint256 threshold) {
+
+    requireSocialRecoveryModuleEnabled();
+
+    require e.msg.value == 0;
+    require threshold > 0;
+    require guardianStorageContract.entries[safeContract].count < max_uint256; // To prevent overflow (Realistically can't reach).
+    require guardianStorageContract.entries[safeContract].count == guardianStorageContract.countGuardians(safeContract);
+
+    require guardian != 0;
+    require guardian != SENTINEL();
+    require guardian != safeContract;
+
+    require !safeContract.isOwner(guardian);
+    require !currentContract.isGuardian(safeContract, guardian);
+
+    require threshold < guardianStorageContract.countGuardians(safeContract);
+
+    require e.msg.sender == safeContract;
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert !isReverted && currentContract.isGuardian(safeContract, guardian);
+}
+
+// This integrity rule verifies that the addition of a new guardian always reverts if the guardian is already added.
+rule addGuardiansRevertIfDuplicateGuardian(env e, address guardian, uint256 threshold) {
+
+    require currentContract.isGuardian(safeContract, guardian);
+
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert lastReverted;
+}
+
+// This integrity rule verifies that the revokeGuardianWithThreshold(...) if executes ensures that
+// the Social Recovery Module is enabled, the caller to the Module has to be the Safe Contract, the
+// guardian is revoked from the guardian list, the linked list integrity remains and no other account
+// (guardian or not) is affected.
+rule revokeGuardiansWorksAsExpected(env e, address guardian, address prevGuardian, uint256 threshold, address otherAccount) {
+
+    address nextGuardian = guardianStorageContract.entries[safeContract].guardians[guardian];
+
+    require guardian != otherAccount;
+    bool otherAccountIsGuardian = currentContract.isGuardian(safeContract, otherAccount);
+
+    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;
+    require currentGuardiansCount > 0;
+
+    require guardianStorageContract.entries[safeContract].count == guardianStorageContract.countGuardians(safeContract);
+    require guardianStorageContract.entries[safeContract].guardians[guardian] != guardian;
```

Comment:
I do not think this require should be here.

I imagine it found a counter example when you had a cyclic list (probably because `count == 0` was checked in the `countGuardians` function and shouldn't be.

```suggestion
```

Reply:
If this is removed, then they errors with next, current and prev guardians as the same address: https://prover.certora.com/output/950385/2db7a0d5601b4d5693cae4b28fa422c1?anonymousKey=ead087388010eb6584a32ac6f1edff53b1dde172

Reply:
The only valid case (which I can think of) which can be avoided by this `require` is when `guardian[SENTINEL] = SENTINEL`. But as we check the count, and ensure that the count is greater than 0, that case is also avoided. Right?

Reply:
Maybe we can take a look together. But I really don't think this require can be there.

In particular, this `guardian[SENTINEL] = SENTINEL` is a valid state and `require`-ing it away would potentially hide bugs that happen in that valid state. If the rule is broken this can mean one or two things:
- The code has some assumption that we are not upholding in the rule
- There is a valid counter example that implies a bug

We should not be adding `require`s just to make CVL rules pass.

Reply:
Looking at the run, it is failing because of this:

![image](https://github.com/5afe/CandideWalletContracts/assets/4210206/dc86779b-c1b1-434b-98b1-81b6af5c454f)

I.e. there is a "ghost" guardian (that is, a guardian that has a non-0 value, but is not part of the linked list). This violates the integrity of the linked list (i.e. `isGuardian(address) => address ∈ getGuardians()`) which is the assumption that we need to `require` for the rule to pass.

Reply:
Removed that `require` and added a check based on the guardian list based on your suggestion 👍🏾 in e8a09f1